### PR TITLE
refactor and export hijack.go

### PR DESCRIPTION
**Pull request by AntaresS**
Signed-off-by: Anda Xu <anda.xu@docker.com>

**- What I did**
- Refactored and exported `hijack.go` code so that other packages can use
- Updated other container code that uses `hijack.go` accordingly

**- How I did it**
- Exported `Stream` function as main entry point

**- How to verify it**
- Run `make -f docker.MakeFile shell` and verified no behavior has been changed

**- Description for the changelog**

Exported `Stream` function in hijack.go so that developers can use it to handle interactive tty.

**- A picture of a cute animal (not mandatory but encouraged)**
image

### CODE REVIEW

**Comment by AntaresS** [comment]
@tiborvass @dnephin Hey as we discussed earlier. Although I still feel like `hijack.go` is still kind of "heavy", at least we can use it freely for other packages now.

**Comment by AntaresS** [comment]
@jlhawn cc you as well, since I've made many changes on your early refactoring.

File: cli/command/container/hijack.go [hunk 1]
```diff
@@ -19,41 +19,41 @@ import (
...
 // connection.
-type hijackedIOStreamer struct {
-	streams      command.Streams
-	inputStream  io.ReadCloser
-	outputStream io.Writer
-	errorStream  io.Writer
-
-	resp types.HijackedResponse
+type HijackOptions struct {
+	//config       *container.Config
```
Comment (tiborvass) [inline cli/command/container/hijack.go]: @AntaresS spurious comment

Comment (AntaresS) [reply cli/command/container/hijack.go]: Oops.. deleted

**Comment by AntaresS** [comment]
@tiborvass not sure why linter is failing...

**Comment by dnephin** [comment]
I clicked rebuild on `lint` and it looks fine now. I'll take a look over the code soon

File: cli/command/container/hijack.go [hunk 2]
```diff
@@ -19,41 +19,40 @@ import (
...
+type HijackOptions struct {
+	resp       types.HijackedResponse
+	ttyInput   io.ReadCloser
+	inputFlag  bool
+	outputFlag bool
+	errorFlag  bool
 
 	tty        bool
 	detachKeys string
 }
```
Comment (jlhawn) [inline cli/command/container/hijack.go]: Exporting this type isn't going to be very helpful unless you also export all of the struct fields 😉

File: cli/command/container/hijack.go [hunk 3]
```diff
@@ -19,41 +19,40 @@ import (
...
-type hijackedIOStreamer struct {
-	streams      command.Streams
-	inputStream  io.ReadCloser
-	outputStream io.Writer
-	errorStream  io.Writer
-
-	resp types.HijackedResponse
+type HijackOptions struct {
+	resp       types.HijackedResponse
+	ttyInput   io.ReadCloser
```
Comment (jlhawn) [inline cli/command/container/hijack.go]: I don't think we necessarily need this field either. Check to see if we actually ever call `Close()` on this value. We might be able to get away removing it the same you did with `outputStream` and `errorStream` and just use the passed in `streams.In()`.

File: cli/command/container/hijack.go [hunk 4]
```diff
@@ -19,41 +19,40 @@ import (
...
+	inputFlag  bool
+	outputFlag bool
+	errorFlag  bool
```
Comment (jlhawn) [inline cli/command/container/hijack.go]: We should be consistent with these names from other struct types: i.e., use `AttachStdin`, `AttachStdout`, and `AttachStderr`.

File: cli/command/container/hijack.go (hunk 3 above)
Comment (AntaresS) [reply cli/command/container/hijack.go]: @jlhawn I saw this io.ReadCloser is tricker than the other two. On [line 104](https://github.com/AntaresS/cli/blob/64f436f08c10656745ce8c489d1f218b3a6f7690/cli/command/container/hijack.go#L104) it is being overwritten and `Close()` is called on [line 202](https://github.com/AntaresS/cli/blob/64f436f08c10656745ce8c489d1f218b3a6f7690/cli/command/container/hijack.go#L202). Seems like it is no longer an on/off switch any more like `outputStream` and `errorStream`

File: cli/command/container/hijack.go (hunk 4 above)
Comment (AntaresS) [reply cli/command/container/hijack.go]: I hesitated on renaming this. Thanks for helping me make the decision ;)

File: cli/command/container/hijack.go (hunk 2 above)
Comment (AntaresS) [reply cli/command/container/hijack.go]: Ah good catch. Forgot to capitalize the fields as well.

**Comment by AntaresS** [comment]
Updated and rebased

File: cli/command/container/hijack.go (hunk 3 above)
Comment (jlhawn) [reply cli/command/container/hijack.go]: The Close() on line 202 is effectively calling `streams.In().Close()` though.

**Comment by AntaresS** [comment]
@dnephin Made a couple of changes accordingly. Please let me know if we can merge it. 🙂

File: cli/command/container/hijack.go [hunk 5]
```diff
@@ -19,41 +19,38 @@ import (
 // TODO: This could be moved to `pkg/term`.
 var defaultEscapeKeys = []byte{16, 17}
 
-// A hijackedIOStreamer handles copying input to and output from streams to the
```
Comment (dnephin) [inline cli/command/container/hijack.go]: This docstring need to be updated to:

"HijackOptions used by Stream to configure streaming behaviour"

File: cli/command/container/hijack.go [hunk 6]
```diff
@@ -71,45 +68,44 @@ func (h *hijackedIOStreamer) stream(ctx context.Context) error {
 	}
 }
 
-func (h *hijackedIOStreamer) setupInput() (restore func(), err error) {
-	if h.inputStream == nil || !h.tty {
+func setupInput(streams command.Streams, opts *HijackOptions) (restore func(), reader io.Reader, err error) {
```
Comment (dnephin) [inline cli/command/container/hijack.go]: I think this should receive a non-pointer type for `opts`. It shouldn't be updating `opts`

File: cli/command/container/hijack.go [hunk 7]
```diff
@@ -19,41 +19,38 @@ import (
...
-	inputStream  io.ReadCloser
-	outputStream io.Writer
-	errorStream  io.Writer
-
-	resp types.HijackedResponse
-
-	tty        bool
-	detachKeys string
+type HijackOptions struct {
+	Resp         types.HijackedResponse
```
Comment (dnephin) [inline cli/command/container/hijack.go]: Maybe `Hijacked` instead of `Resp` for the field name?

Comment (AntaresS) [reply cli/command/container/hijack.go]: I don't hold strong opinion on naming this guy. Changed to `Hijacked`

**Comment by AntaresS** [comment]
> I think we should move hijack.go to a new package, so consumers don't have to include all the dependencies of cli/command/container.
> A new package at cli/hijack/ might be good.
> Another problem is we have basically no tests for this right now. Some test coverage would be great. It looks like the build is failing right now.
@dnephin @tiborvass 
I moved `hijack.go` to `cli/hijack` folder. Don't think this part of code should belong to container cli either. 
For the unit test, is it a must requirement for this PR? Otherwise I suggest we ship unit test in another PR. I hope we can get this one merged in order to unblock my other task 🙂

**Comment by dnephin** [comment]
#495 should provide some indirect converage, but some a couple unit tests would really be good.

Looks like there is one lint error.

You can use `vndr` to point at your branch so that you can start developing against this before it gets merged.

**Comment by thaJeztah** [comment]
ping @AntaresS are you still working on this? Looks like this needs a rebase now

**Comment by thaJeztah** [comment]
ping @AntaresS 😇

**Comment by AntaresS** [comment]
@thaJeztah Sorry I lost track on this PR. Let's close it since the original feature request in testkit is no longer active.

**Pull request closed without merging**